File: Downloads/ai_development_mass_framework-20250622T063827Z-1-001/ai_development_mass_framework/core/cloud_deployment_service.py
"""
Cloud Deployment Service for MASS Framework
Handles deployment to various cloud providers including Docker, Kubernetes, AWS, etc.
"""

import json
import secrets
import asyncio
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass, asdict
from enum import Enum
import logging
import subprocess
import os

logger = logging.getLogger(__name__)

class CloudProvider(Enum):
    DOCKER = "docker"
    KUBERNETES = "kubernetes"
    AWS = "aws"
    GCP = "gcp"
    AZURE = "azure"

class DeploymentStatus(Enum):
    PENDING = "pending"
    DEPLOYING = "deploying"
    DEPLOYED = "deployed"
    FAILED = "failed"
    TERMINATED = "terminated"

class ScalingEventType(Enum):
    SCALE_UP = "scale_up"
    SCALE_DOWN = "scale_down"
    AUTO_SCALE = "auto_scale"

@dataclass
class ScalingConfig:
    min_replicas: int = 1
    max_replicas: int = 3
    target_cpu_utilization: int = 70
    memory_limit: str = "512Mi"

@dataclass
class SecurityConfig:
    enable_https: bool = False
    enable_authentication: bool = True
    enable_rate_limiting: bool = True
    api_keys_required: bool = True

@dataclass
class MonitoringConfig:
    enable_metrics: bool = True
    enable_logging: bool = True
    enable_alerting: bool = False
    log_level: str = "info"

@dataclass
class DeploymentConfig:
    environment: str
    cloud_provider: CloudProvider
    scaling_config: ScalingConfig
    security_config: SecurityConfig
    monitoring_config: MonitoringConfig

@dataclass
class ResourceMetrics:
    cpu_usage: float = 0.0
    memory_usage: float = 0.0
    active_requests: int = 0
    response_time_avg: float = 0.0

@dataclass
class HealthChecks:
    api_health: bool = True
    database_health: bool = True
    ai_services_health: bool = True

@dataclass
class DeploymentEndpoints:
    api_url: str
    admin_url: Optional[str] = None
    monitoring_url: Optional[str] = None

@dataclass
class Deployment:
    deployment_id: str
    status: DeploymentStatus
    config: DeploymentConfig
    created_at: datetime
    updated_at: datetime
    endpoints: DeploymentEndpoints
    resources: ResourceMetrics
    health_checks: HealthChecks
    replicas: int = 1

@dataclass
class ScalingEvent:
    timestamp: datetime
    event_type: ScalingEventType
    from_replicas: int
    to_replicas: int
    reason: str
    triggered_by: str

class CloudDeploymentService:
    """Service for managing cloud deployments of the MASS Framework"""
    
    def __init__(self):
        self.deployments: Dict[str, Deployment] = {}
        self.scaling_history: Dict[str, List[ScalingEvent]] = {}
        self._load_existing_deployments()
    
    def _load_existing_deployments(self):
        """Load existing deployments from persistent storage"""
        # In a real implementation, this would load from a database
        # For now, we'll start with empty state
        pass
    
    async def deploy_to_cloud(self, config: DeploymentConfig) -> Dict[str, Any]:
        """Deploy the MASS Framework to the specified cloud provider"""
        deployment_id = f"mass-{secrets.token_urlsafe(8)}"
        
        try:
            # Create deployment record
            deployment = Deployment(
                deployment_id=deployment_id,
                status=DeploymentStatus.DEPLOYING,
                config=config,
                created_at=datetime.now(),
                updated_at=datetime.now(),
                endpoints=DeploymentEndpoints(api_url=""),
                resources=ResourceMetrics(),
                health_checks=HealthChecks(),
                replicas=config.scaling_config.min_replicas
            )
            
            self.deployments[deployment_id] = deployment
            self.scaling_history[deployment_id] = []
            
            # Perform actual deployment based on provider
            success = await self._deploy_to_provider(deployment)
            
            if success:
                deployment.status = DeploymentStatus.DEPLOYED
                deployment.endpoints.api_url = f"http://mass-{deployment_id}.example.com"
                if config.environment == "production":
                    deployment.endpoints.admin_url = f"http://admin-{deployment_id}.example.com"
                    deployment.endpoints.monitoring_url = f"http://monitoring-{deployment_id}.example.com"
                
                # Record initial scaling event
                self.scaling_history[deployment_id].append(
                    ScalingEvent(
                        timestamp=datetime.now(),
                        event_type=ScalingEventType.SCALE_UP,
                        from_replicas=0,
                        to_replicas=deployment.replicas,
                        reason="Initial deployment",
                        triggered_by="deployment"
                    )
                )
                
                logger.info(f"Deployment {deployment_id} completed successfully")
            else:
                deployment.status = DeploymentStatus.FAILED
                logger.error(f"Deployment {deployment_id} failed")
            
            deployment.updated_at = datetime.now()
            
            return {
                "deployment_id": deployment_id,
                "status": deployment.status.value,
                "endpoints": asdict(deployment.endpoints) if success else {}
            }
            
        except Exception as e:
            logger.error(f"Deployment failed: {str(e)}")
            if deployment_id in self.deployments:
                self.deployments[deployment_id].status = DeploymentStatus.FAILED
            raise
    
    async def _deploy_to_provider(self, deployment: Deployment) -> bool:
        """Deploy to the specific cloud provider"""
        provider = deployment.config.cloud_provider
        
        try:
            if provider == CloudProvider.DOCKER:
                return await self._deploy_docker(deployment)
            elif provider == CloudProvider.KUBERNETES:
                return await self._deploy_kubernetes(deployment)
            elif provider == CloudProvider.AWS:
                return await self._deploy_aws(deployment)
            elif provider == CloudProvider.GCP:
                return await self._deploy_gcp(deployment)
            elif provider == CloudProvider.AZURE:
                return await self._deploy_azure(deployment)
            else:
                logger.error(f"Unsupported cloud provider: {provider}")
                return False
                
        except Exception as e:
            logger.error(f"Provider deployment failed: {str(e)}")
            return False
    
    async def _deploy_docker(self, deployment: Deployment) -> bool:
        """Deploy using Docker"""
        try:
            # Build and run Docker containers
            deployment_name = f"mass-{deployment.deployment_id}"
            
            # Generate docker-compose configuration
            compose_config = self._generate_docker_compose(deployment)
            
            # Write compose file
            compose_file = f"/tmp/{deployment_name}-compose.yml"
            with open(compose_file, 'w') as f:
                f.write(compose_config)
            
            # Deploy using docker-compose
            result = subprocess.run([
                "docker-compose", "-f", compose_file, "-p", deployment_name, "up", "-d"
            ], capture_output=True, text=True)
            
            if result.returncode == 0:
                logger.info(f"Docker deployment {deployment_name} started successfully")
                return True
            else:
                logger.error(f"Docker deployment failed: {result.stderr}")
                return False
                
        except Exception as e:
            logger.error(f"Docker deployment error: {str(e)}")
            return False
    
    async def _deploy_kubernetes(self, deployment: Deployment) -> bool:
        """Deploy using Kubernetes"""
        try:
            # Generate Kubernetes manifests
            manifests = self._generate_k8s_manifests(deployment)
            
            # Apply manifests
            for manifest_name, manifest_content in manifests.items():
                manifest_file = f"/tmp/{deployment.deployment_id}-{manifest_name}.yaml"
                with open(manifest_file, 'w') as f:
                    f.write(manifest_content)
                
                result = subprocess.run([
                    "kubectl", "apply", "-f", manifest_file
                ], capture_output=True, text=True)
                
                if result.returncode != 0:
                    logger.error(f"Failed to apply {manifest_name}: {result.stderr}")
                    return False
            
            logger.info(f"Kubernetes deployment {deployment.deployment_id} completed")
            return True
            
        except Exception as e:
            logger.error(f"Kubernetes deployment error: {str(e)}")
            return False
    
    async def _deploy_aws(self, deployment: Deployment) -> bool:
        """Deploy using AWS services (ECS, EKS, etc.)"""
        # Placeholder for AWS deployment
        logger.info("AWS deployment not yet implemented")
        return True
    
    async def _deploy_gcp(self, deployment: Deployment) -> bool:
        """Deploy using Google Cloud services"""
        # Placeholder for GCP deployment
        logger.info("GCP deployment not yet implemented")
        return True
    
    async def _deploy_azure(self, deployment: Deployment) -> bool:
        """Deploy using Azure services"""
        # Placeholder for Azure deployment
        logger.info("Azure deployment not yet implemented")
        return True
    
    def _generate_docker_compose(self, deployment: Deployment) -> str:
        """Generate docker-compose.yml for deployment"""
        config = deployment.config
        
        return f"""
version: '3.8'
services:
  mass-backend:
    image: mass-framework:latest
    ports:
      - "8000:8000"
    environment:
      - ENVIRONMENT={config.environment}
      - ENABLE_AUTH={config.security_config.enable_authentication}
      - ENABLE_RATE_LIMITING={config.security_config.enable_rate_limiting}
      - LOG_LEVEL={config.monitoring_config.log_level}
    deploy:
      replicas: {deployment.replicas}
      resources:
        limits:
          memory: {config.scaling_config.memory_limit}
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8000/health"]
      interval: 30s
      timeout: 10s
      retries: 3
      
  mass-frontend:
    image: mass-frontend:latest
    ports:
      - "3000:80"
    depends_on:
      - mass-backend
    deploy:
      replicas: 1
"""
    
    def _generate_k8s_manifests(self, deployment: Deployment) -> Dict[str, str]:
        """Generate Kubernetes manifests for deployment"""
        config = deployment.config
        
        deployment_manifest = f"""
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mass-backend-{deployment.deployment_id}
  labels:
    app: mass-backend
    deployment-id: {deployment.deployment_id}
spec:
  replicas: {deployment.replicas}
  selector:
    matchLabels:
      app: mass-backend
      deployment-id: {deployment.deployment_id}
  template:
    metadata:
      labels:
        app: mass-backend
        deployment-id: {deployment.deployment_id}
    spec:
      containers:
      - name: mass-backend
        image: mass-framework:latest
        ports:
        - containerPort: 8000
        env:
        - name: ENVIRONMENT
          value: "{config.environment}"
        - name: ENABLE_AUTH
          value: "{config.security_config.enable_authentication}"
        - name: LOG_LEVEL
          value: "{config.monitoring_config.log_level}"
        resources:
          limits:
            memory: {config.scaling_config.memory_limit}
          requests:
            memory: "256Mi"
        livenessProbe:
          httpGet:
            path: /health
            port: 8000
          initialDelaySeconds: 30
          periodSeconds: 10
"""
        
        service_manifest = f"""
apiVersion: v1
kind: Service
metadata:
  name: mass-backend-service-{deployment.deployment_id}
spec:
  selector:
    app: mass-backend
    deployment-id: {deployment.deployment_id}
  ports:
  - protocol: TCP
    port: 80
    targetPort: 8000
  type: LoadBalancer
"""
        
        if config.scaling_config.max_replicas > config.scaling_config.min_replicas:
            hpa_manifest = f"""
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: mass-backend-hpa-{deployment.deployment_id}
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: mass-backend-{deployment.deployment_id}
  minReplicas: {config.scaling_config.min_replicas}
  maxReplicas: {config.scaling_config.max_replicas}
  metrics:
  - type: Resource
    resource:
      name: cpu
      target:
        type: Utilization
        averageUtilization: {config.scaling_config.target_cpu_utilization}
"""
        else:
            hpa_manifest = ""
        
        manifests = {
            "deployment": deployment_manifest,
            "service": service_manifest
        }
        
        if hpa_manifest:
            manifests["hpa"] = hpa_manifest
        
        return manifests
    
    async def get_deployment_status(self, deployment_id: str) -> Optional[Dict[str, Any]]:
        """Get the status of a specific deployment"""
        if deployment_id not in self.deployments:
            return None
        
        deployment = self.deployments[deployment_id]
        
        # Update resource metrics (in real implementation, this would fetch from monitoring)
        if deployment.status == DeploymentStatus.DEPLOYED:
            await self._update_resource_metrics(deployment)
        
        return {
            "deployment_id": deployment.deployment_id,
            "status": deployment.status.value,
            "created_at": deployment.created_at.isoformat(),
            "updated_at": deployment.updated_at.isoformat(),
            "endpoints": asdict(deployment.endpoints),
            "resources": asdict(deployment.resources),
            "health_checks": asdict(deployment.health_checks)
        }
    
    async def list_deployments(self) -> List[Dict[str, Any]]:
        """List all deployments"""
        deployments = []
        for deployment in self.deployments.values():
            if deployment.status == DeploymentStatus.DEPLOYED:
                await self._update_resource_metrics(deployment)
            deployments.append(await self.get_deployment_status(deployment.deployment_id))
        
        return [d for d in deployments if d is not None]
    
    async def scale_deployment(self, deployment_id: str, replicas: int, reason: str = "Manual scaling", triggered_by: str = "manual") -> bool:
        """Scale a deployment to the specified number of replicas"""
        if deployment_id not in self.deployments:
            raise ValueError(f"Deployment {deployment_id} not found")
        
        deployment = self.deployments[deployment_id]
        
        if deployment.status != DeploymentStatus.DEPLOYED:
            raise ValueError(f"Cannot scale deployment in status: {deployment.status.value}")
        
        old_replicas = deployment.replicas
        deployment.replicas = replicas
        deployment.updated_at = datetime.now()
        
        # Perform actual scaling based on provider
        success = await self._scale_provider_deployment(deployment)
        
        if success:
            # Record scaling event
            event_type = ScalingEventType.SCALE_UP if replicas > old_replicas else ScalingEventType.SCALE_DOWN
            self.scaling_history[deployment_id].append(
                ScalingEvent(
                    timestamp=datetime.now(),
                    event_type=event_type,
                    from_replicas=old_replicas,
                    to_replicas=replicas,
                    reason=reason,
                    triggered_by=triggered_by
                )
            )
            logger.info(f"Scaled deployment {deployment_id} from {old_replicas} to {replicas} replicas")
        else:
            # Revert on failure
            deployment.replicas = old_replicas
            logger.error(f"Failed to scale deployment {deployment_id}")
        
        return success
    
    async def _scale_provider_deployment(self, deployment: Deployment) -> bool:
        """Scale the deployment using the specific provider"""
        provider = deployment.config.cloud_provider
        
        try:
            if provider == CloudProvider.DOCKER:
                return await self._scale_docker(deployment)
            elif provider == CloudProvider.KUBERNETES:
                return await self._scale_kubernetes(deployment)
            # Add other providers as needed
            else:
                logger.warning(f"Scaling not implemented for provider: {provider}")
                return True  # Mock success for now
                
        except Exception as e:
            logger.error(f"Scaling failed: {str(e)}")
            return False
    
    async def _scale_docker(self, deployment: Deployment) -> bool:
        """Scale Docker deployment"""
        try:
            # Use docker-compose scale command
            deployment_name = f"mass-{deployment.deployment_id}"
            result = subprocess.run([
                "docker-compose", "-p", deployment_name, "scale", 
                f"mass-backend={deployment.replicas}"
            ], capture_output=True, text=True)
            
            return result.returncode == 0
        except Exception as e:
            logger.error(f"Docker scaling error: {str(e)}")
            return False
    
    async def _scale_kubernetes(self, deployment: Deployment) -> bool:
        """Scale Kubernetes deployment"""
        try:
            result = subprocess.run([
                "kubectl", "scale", "deployment", 
                f"mass-backend-{deployment.deployment_id}",
                f"--replicas={deployment.replicas}"
            ], capture_output=True, text=True)
            
            return result.returncode == 0
        except Exception as e:
            logger.error(f"Kubernetes scaling error: {str(e)}")
            return False
    
    async def terminate_deployment(self, deployment_id: str) -> bool:
        """Terminate a deployment"""
        if deployment_id not in self.deployments:
            raise ValueError(f"Deployment {deployment_id} not found")
        
        deployment = self.deployments[deployment_id]
        
        success = await self._terminate_provider_deployment(deployment)
        
        if success:
            deployment.status = DeploymentStatus.TERMINATED
            deployment.updated_at = datetime.now()
            logger.info(f"Terminated deployment {deployment_id}")
        else:
            logger.error(f"Failed to terminate deployment {deployment_id}")
        
        return success
    
    async def _terminate_provider_deployment(self, deployment: Deployment) -> bool:
        """Terminate deployment using the specific provider"""
        provider = deployment.config.cloud_provider
        
        try:
            if provider == CloudProvider.DOCKER:
                return await self._terminate_docker(deployment)
            elif provider == CloudProvider.KUBERNETES:
                return await self._terminate_kubernetes(deployment)
            else:
                logger.warning(f"Termination not implemented for provider: {provider}")
                return True  # Mock success for now
                
        except Exception as e:
            logger.error(f"Termination failed: {str(e)}")
            return False
    
    async def _terminate_docker(self, deployment: Deployment) -> bool:
        """Terminate Docker deployment"""
        try:
            deployment_name = f"mass-{deployment.deployment_id}"
            result = subprocess.run([
                "docker-compose", "-p", deployment_name, "down"
            ], capture_output=True, text=True)
            
            return result.returncode == 0
        except Exception as e:
            logger.error(f"Docker termination error: {str(e)}")
            return False
    
    async def _terminate_kubernetes(self, deployment: Deployment) -> bool:
        """Terminate Kubernetes deployment"""
        try:
            # Delete all resources for this deployment
            result = subprocess.run([
                "kubectl", "delete", "all", "-l", 
                f"deployment-id={deployment.deployment_id}"
            ], capture_output=True, text=True)
            
            return result.returncode == 0
        except Exception as e:
            logger.error(f"Kubernetes termination error: {str(e)}")
            return False
    
    async def _update_resource_metrics(self, deployment: Deployment):
        """Update resource metrics for a deployment"""
        # In a real implementation, this would fetch metrics from monitoring systems
        # For now, we'll simulate some metrics
        import random
        
        deployment.resources.cpu_usage = random.uniform(20, 80)
        deployment.resources.memory_usage = random.uniform(30, 70)
        deployment.resources.active_requests = random.randint(0, 50)
        deployment.resources.response_time_avg = random.uniform(50, 200)
    
    async def get_metrics(self, deployment_id: str, time_range: str = "1h") -> Dict[str, Any]:
        """Get metrics for a deployment"""
        if deployment_id not in self.deployments:
            raise ValueError(f"Deployment {deployment_id} not found")
        
        # Generate mock metrics data
        now = datetime.now()
        metrics = {
            "cpu_metrics": [],
            "memory_metrics": [],
            "request_metrics": [],
            "error_metrics": []
        }
        
        # Generate data points based on time range
        points = 20 if time_range == "1h" else 100
        interval = timedelta(hours=1) / points if time_range == "1h" else timedelta(hours=24) / points
        
        import random
        for i in range(points):
            timestamp = now - (interval * (points - i))
            
            metrics["cpu_metrics"].append({
                "timestamp": timestamp.isoformat(),
                "value": random.uniform(20, 80)
            })
            
            metrics["memory_metrics"].append({
                "timestamp": timestamp.isoformat(),
                "value": random.uniform(30, 70)
            })
            
            metrics["request_metrics"].append({
                "timestamp": timestamp.isoformat(),
                "requests_per_second": random.uniform(5, 50),
                "avg_response_time": random.uniform(50, 200)
            })
            
            metrics["error_metrics"].append({
                "timestamp": timestamp.isoformat(),
                "error_rate": random.uniform(0, 5)
            })
        
        return metrics
    
    def get_scaling_history(self, deployment_id: str) -> List[Dict[str, Any]]:
        """Get scaling history for a deployment"""
        if deployment_id not in self.scaling_history:
            return []
        
        return [
            {
                "timestamp": event.timestamp.isoformat(),
                "event_type": event.event_type.value,
                "from_replicas": event.from_replicas,
                "to_replicas": event.to_replicas,
                "reason": event.reason,
                "triggered_by": event.triggered_by
            }
            for event in self.scaling_history[deployment_id]
        ]
    
    def get_deployment_templates(self) -> Dict[str, DeploymentConfig]:
        """Get predefined deployment configuration templates"""
        return {
            "development": DeploymentConfig(
                environment="development",
                cloud_provider=CloudProvider.DOCKER,
                scaling_config=ScalingConfig(
                    min_replicas=1,
                    max_replicas=2,
                    target_cpu_utilization=80,
                    memory_limit="512Mi"
                ),
                security_config=SecurityConfig(
                    enable_https=False,
                    enable_authentication=True,
                    enable_rate_limiting=False,
                    api_keys_required=False
                ),
                monitoring_config=MonitoringConfig(
                    enable_metrics=True,
                    enable_logging=True,
                    enable_alerting=False,
                    log_level="debug"
                )
            ),
            "staging": DeploymentConfig(
                environment="staging",
                cloud_provider=CloudProvider.KUBERNETES,
                scaling_config=ScalingConfig(
                    min_replicas=2,
                    max_replicas=5,
                    target_cpu_utilization=70,
                    memory_limit="1Gi"
                ),
                security_config=SecurityConfig(
                    enable_https=True,
                    enable_authentication=True,
                    enable_rate_limiting=True,
                    api_keys_required=True
                ),
                monitoring_config=MonitoringConfig(
                    enable_metrics=True,
                    enable_logging=True,
                    enable_alerting=True,
                    log_level="info"
                )
            ),
            "production": DeploymentConfig(
                environment="production",
                cloud_provider=CloudProvider.KUBERNETES,
                scaling_config=ScalingConfig(
                    min_replicas=3,
                    max_replicas=10,
                    target_cpu_utilization=60,
                    memory_limit="2Gi"
                ),
                security_config=SecurityConfig(
                    enable_https=True,
                    enable_authentication=True,
                    enable_rate_limiting=True,
                    api_keys_required=True
                ),
                monitoring_config=MonitoringConfig(
                    enable_metrics=True,
                    enable_logging=True,
                    enable_alerting=True,
                    log_level="warning"
                )
            )
        }

# Global instance
cloud_deployment_service = CloudDeploymentService()

File: Downloads/ai_development_mass_framework-20250622T063827Z-1-001/ai_development_mass_framework/frontend/src/services/cloudDeploymentService.ts
/**
 * Cloud Deployment Service for MASS Framework
 * Handles cloud deployment, scaling, and monitoring operations
 */

import enhancedApiService from './enhancedApiService';

// Demo mode flag - set to true for demo/development
const DEMO_MODE = true;
const DEMO_API_BASE = DEMO_MODE ? '/api/demo/cloud' : '/api/cloud';

export interface DeploymentConfig {
  environment: 'development' | 'staging' | 'production';
  cloud_provider: 'aws' | 'gcp' | 'azure' | 'docker' | 'kubernetes';
  scaling_config: {
    min_replicas: number;
    max_replicas: number;
    target_cpu_utilization: number;
    memory_limit: string;
  };
  security_config: {
    enable_https: boolean;
    enable_authentication: boolean;
    enable_rate_limiting: boolean;
    api_keys_required: boolean;
  };
  monitoring_config: {
    enable_metrics: boolean;
    enable_logging: boolean;
    enable_alerting: boolean;
    log_level: 'debug' | 'info' | 'warning' | 'error';
  };
}

export interface DeploymentStatus {
  deployment_id: string;
  status: 'pending' | 'deploying' | 'deployed' | 'failed' | 'terminated';
  created_at: string;
  updated_at: string;
  endpoints: {
    api_url: string;
    admin_url?: string;
    monitoring_url?: string;
  };
  resources: {
    cpu_usage: number;
    memory_usage: number;
    active_requests: number;
    response_time_avg: number;
  };
  health_checks: {
    api_health: boolean;
    database_health: boolean;
    ai_services_health: boolean;
  };
}

export interface ScalingEvent {
  timestamp: string;
  event_type: 'scale_up' | 'scale_down' | 'auto_scale';
  from_replicas: number;
  to_replicas: number;
  reason: string;
  triggered_by: 'manual' | 'cpu_threshold' | 'memory_threshold' | 'request_load';
}

class CloudDeploymentService {
  private baseUrl: string;
  private apiKey: string | null = null;

  constructor() {
    this.baseUrl = process.env.REACT_APP_API_URL || 'http://localhost:8000';
  }

  setApiKey(apiKey: string) {
    this.apiKey = apiKey;
  }

  private getHeaders(): HeadersInit {
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
    };
    
    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }
    
    return headers;
  }
  /**
   * Deploy the MASS Framework to cloud
   */
  async deployToCloud(config: DeploymentConfig & { app_name: string }): Promise<{ deployment_id: string; status: string }> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/deploy`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(config)
    });

    if (!response.ok) {
      throw new Error(`Deployment failed: ${response.statusText}`);
    }

    return response.json();
  }
  /**
   * Get deployment status and metrics
   */
  async getDeploymentStatus(deploymentId: string): Promise<DeploymentStatus> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/status/${deploymentId}`, {
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch deployment status: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * List all active deployments
   */
  async listDeployments(): Promise<DeploymentStatus[]> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/deployments`, {
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch deployments: ${response.statusText}`);
    }

    const data = await response.json();
    return data.deployments || [];
  }

  /**
   * Scale deployment replicas
   */
  async scaleDeployment(deploymentId: string, replicas: number, reason?: string): Promise<{ success: boolean }> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/deployments/${deploymentId}/scale`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ 
        replicas, 
        reason: reason || 'Manual scaling',
        triggered_by: 'manual'
      })
    });

    if (!response.ok) {
      throw new Error(`Scaling failed: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get scaling history
   */
  async getScalingHistory(deploymentId: string): Promise<ScalingEvent[]> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/deployments/${deploymentId}/scaling-history`, {
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch scaling history: ${response.statusText}`);
    }

    const data = await response.json();
    return data.events || [];
  }

  /**
   * Terminate deployment
   */
  async terminateDeployment(deploymentId: string): Promise<{ success: boolean }> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/deployments/${deploymentId}/terminate`, {
      method: 'DELETE',
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Termination failed: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get real-time metrics
   */
  async getMetrics(deploymentId: string, timeRange: '1h' | '6h' | '24h' | '7d' = '1h'): Promise<{
    cpu_metrics: Array<{ timestamp: string; value: number }>;
    memory_metrics: Array<{ timestamp: string; value: number }>;
    request_metrics: Array<{ timestamp: string; requests_per_second: number; avg_response_time: number }>;
    error_metrics: Array<{ timestamp: string; error_rate: number }>;
  }> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/deployments/${deploymentId}/metrics?range=${timeRange}`, {
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch metrics: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get available cloud providers
   */
  async getCloudProviders(): Promise<{ providers: Array<{ id: string; name: string; description: string; supported_services: string[]; status: string }> }> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/providers`, {
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch cloud providers: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get deployment templates
   */
  async getDeploymentTemplates(): Promise<{ templates: Array<{ id: string; name: string; description: string; cloud_providers: string[]; config: any }> }> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/templates`, {
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch deployment templates: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get deployment metrics
   */
  async getDeploymentMetrics(deploymentId: string): Promise<{ deployment_id: string; metrics: any[]; summary: any }> {
    const response = await fetch(`${this.baseUrl}${DEMO_API_BASE}/metrics/${deploymentId}`, {
      headers: this.getHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch deployment metrics: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Monitor deployment health in real-time
   */
  createHealthMonitor(deploymentId: string, callback: (health: DeploymentStatus['health_checks']) => void): WebSocket | null {
    try {
      const wsUrl = this.baseUrl.replace('http', 'ws') + `${DEMO_API_BASE}/deployments/${deploymentId}/health-monitor`;
      const ws = new WebSocket(wsUrl);

      ws.onmessage = (event) => {
        try {
          const health = JSON.parse(event.data);
          callback(health);
        } catch (error) {
          console.error('Failed to parse health data:', error);
        }
      };

      ws.onerror = (error) => {
        console.error('Health monitor WebSocket error:', error);
      };

      return ws;
    } catch (error) {
      console.error('Failed to create health monitor:', error);
      return null;
    }
  }
}

export const cloudDeploymentService = new CloudDeploymentService();
export default cloudDeploymentService;
